src/ApplicationInsights.Kubernetes/K8sHttpClient/KubeHttpClientSettingsBase.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            if (!File.Exists(pathToNamespace))
+            {
+                throw new FileNotFoundException("File contains namespace does not exist.", pathToNamespace);
+            }
+            return File.ReadAllText(pathToNamespace);
+        }
+
+        /// <summary>
+        /// Verify server certificate within returned HttpResponse to prevent MITM attack.
+        /// </summary>
+        /// <param name="httpRequestMessage">The httpRequestMessage returned.</param>
+        /// <param name="serverCertVerifier">The server certificate.</param>
+        /// <param name="chain">The X509Chain.</param>
+        /// <param name="policyErrors">SslPolicyErrors.</param>
+        /// <param name="clientCertVerifier">Client certificate.</param>
+        /// <returns>Returns true when server certificate is valid.</returns>
+        internal bool VerifyServerCertificate(HttpRequestMessage httpRequestMessage, ICertificateVerifier serverCertVerifier, X509Chain chain, SslPolicyErrors policyErrors, ICertificateVerifier clientCertVerifier)
+        {
+            Arguments.IsNotNull(clientCertVerifier, nameof(clientCertVerifier));
+            X509Certificate2 serverCert = serverCertVerifier.Certificate;
+            _logger.LogDebug("Server certification custom validation callback.");
+            _logger.LogTrace(httpRequestMessage?.ToString());
+            _logger.LogTrace(chain?.ToString());
+            _logger.LogTrace(policyErrors.ToString());
+            _logger.LogTrace("ServerCert:" + Environment.NewLine + serverCert);
+
+            try
+            {
+                // Verify Issuer. Issuer field is case-insensitive.
+                if (!string.Equals(clientCertVerifier.Issuer, serverCertVerifier.Issuer, StringComparison.OrdinalIgnoreCase))
+                {
+                    _logger.LogError(Invariant($"Issuer are different for server certificate and the client certificate. Server Certificate Issuer: {clientCertVerifier.Issuer}, Client Certificate Issuer: {serverCertVerifier.Issuer}"));
+                    return false;
+                }
+                else
+                {
+                    _logger.LogDebug(Invariant($"Issuer validation passed: {serverCertVerifier.Issuer}"));
+                }
+
+                // Server certificate is not expired.
+                DateTime now = DateTime.Now;
+                if (serverCertVerifier.NotBefore > now || serverCertVerifier.NotAfter.AddDays(1) <= now)
+                {
+                    _logger.LogError(Invariant($"Server certification is not in valid period from {serverCertVerifier.NotBefore.ToString(DateTimeFormatInfo.InvariantInfo)} until {serverCertVerifier.NotAfter.ToString(DateTimeFormatInfo.InvariantInfo)}"));
+                    return false;
+                }
+                else
+                {
+                    _logger.LogDebug("Server certificate validate date verification passed.");
+                }
```

Comment:
Certificate validation should always check _three_ things:

1. Issued by trusted party
1. Not expired
1. Issued for service we are trying to communicate with

Seems like the third check is missing?

Reply:
By current understand, the implementation of the certification validation is totally wrong.
**Details:**
By current understanding, it should be replaced as in https://github.com/kubernetes-client/csharp/blob/ea62ca00e099571be0e08022c27c44ef5f9392a5/src/KubernetesClient/Kubernetes.ConfigInit.cs
If possible, I would like to replace the whole query library with https://github.com/kubernetes-client/ when feasible.

Reply:
Update: I guess I have a fix. Code will be updated later.